Sources/Camille/CrossPostService.swift

```diff
@@ -0,0 +1,201 @@
+                else { return }
+            
+            try webApi.execute(message.makeChatPostMessage(target: target))
+        }
+    }
+    override func messageEvent(slackBot: SlackBot, webApi: WebAPI, message: MessageDecorator, previous: MessageDecorator?) throws {
+        self.messages.append(message)
+    }
+}
+
+//MARK: - CrossPost Check
+fileprivate extension CrossPostService {
+    func makeCrossPostWarning(pong: Pong, messages: [MessageDecorator], webApi: WebAPI) -> SlackMessage? {
+        let duplicates = self.findCrossPosts(in: messages)
+        guard !duplicates.isEmpty else { return nil }
+        
+        let message = SlackMessage()
+            .line("Cross Post Alert:".bold)
+            .line("The following cross posts have been detected:")
+        
+        duplicates.addAttachments(
+            to: message,
+            buttonResponder: self,
+            handler: self.buttonHandler(messages: messages, webApi: webApi)
+        )
+        
+        return message
+    }
+}
+
+//MARK: - CrossPost Data
+fileprivate extension CrossPostService {
+    func uniqueMessageKey(message: MessageDecorator) -> String {
+        let userId = message.sender?.id ?? ""
+        return "\(userId)\(message.text.hashValue)"
+    }
+    func potentialDuplicates(messages: [MessageDecorator]) -> Bool {
+        guard
+            messages.count > 1,
+            let first = messages.first,
+            first.sender != nil
+            else { return false }
+        
+        return !first.text.isEmpty
+    }
+    func findCrossPosts(in messages: [MessageDecorator]) -> [[MessageDecorator]] {
+        let potentialDuplicates = messages
+            .grouped(by: self.uniqueMessageKey)
+            .values
+            .filter(self.potentialDuplicates)
+        
+        return Array(potentialDuplicates)
+    }
+}
+
+//MARK: - CrossPost Warning Message Generation
+fileprivate enum ActionButton: String {
+    case privateWarning
+    case publicWarning
+    case removeAll
+    
+    var text: String {
+        switch self {
+        case .privateWarning: return "Private Warning"
+        case .publicWarning: return "Public Warning"
+        case .removeAll: return "Remove all posts"
+        }
+    }
+    
+    static var all: [ActionButton] { return [.privateWarning, .publicWarning, .removeAll] }
+}
+fileprivate extension Sequence where Iterator.Element == [MessageDecorator] {
+    func addAttachments(to message: SlackMessage, buttonResponder: SlackInteractiveButtonResponderService, handler: @escaping InteractiveButtonResponseHandler) {
+        var message = message
+        message = self.reduce(message) { current, messages in
+            return current.attachment { builder in
+                messages.addAttachment(
+                    with: builder,
+                    buttonResponder: buttonResponder,
+                    handler: handler
+                )
+            }
+        }
```

Comment:
This one is basically a failure on my part to provide a nicer syntax.. You can only apply attachments 1 at a time...
`var` is required because the value is being reassigned even though its a reference type :(

Now that I'm seeing it again though I think I'm gonna write something a little nicer for this case - stand by 😄